Allow testing of the documentation (#1613)
sherpa/sherpa: docs/developer/index.rst

The report is in ``report/index.html``, which links to individual
files and shows exactly which lines were excuted while running the tests.

Run doctests locally
--------------------
If `doctestplus <https://pypi.org/project/pytest-doctestplus/>` is installed
(and it probably is because it's part of
`sphinx-astropy <https://pypi.org/project/sphinx-astropy/>`,
which is required to build the documentation locally), you can
run the examples in the documentation or in the doctrings of individual

> functions, for example:

Reviewer: ```suggestion
functions, for example::
```

There's a problem with the current documentation since the following `pytest` lines do not get written as separate lines. I assume that adding a colon here will fix it, but this is just a guess...

![screenshot](https://user-images.githubusercontent.com/224638/199530549-226419cd-288e-4090-9162-9c3631ab3e2c.png)

Reviewer: So, I ran

    pytest --doctest-plus

and it collected 7623 tests, which is the same number as I got from just `pytest`. Is this expected?

As a further check,

    pytest --doctest-plus sherpa/astro/data.py
    pytest sherpa/astro/data.py

both run 13 tests. So how do I know that it's actually running the extra tests - do I just have to assume it is?

Author: To answer the first part of this: With the changes in this PR, doctest-plus is automatically activated for all files that are NOT on the doctestplus blacklist. You can see that the number of tests increases compared to main.
Note that the number shown in my output below is different from what @DougBurke said above, but that's OK because (a) the number of tests run depends on the plotting/IO backend and (b) the number of tests has changed in main when I just rebased locally (not pushed yet). The point is that the settings in this PR increas the number of tests compared to main. The increase is 6, because there are 6 files in the rst files in the docs that are **NOT** blacklisted in `pytest.ini` and doctest run each rst file as a single test.
On main
```
(sherpaalone) MoritzAirRoseGold ~/code/sherpa> pytest
================================================================================================== test session starts ==================================================================================================
platform darwin -- Python 3.10.0, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/guenther/code/sherpa, configfile: pytest.ini
plugins: doctestplus-0.11.2, typeguard-2.13.3, xvfb-2.0.0
collecting ... WARNING: Failed to import any usable sherpa.plotting backend. Plotting routines will not be available.
collected 7547 items       
```

On this branch:
```
(sherpaalone) MoritzAirRoseGold ~/code/sherpa> pytest 
================================================================================================== test session starts ==================================================================================================
platform darwin -- Python 3.10.0, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/guenther/code/sherpa, configfile: pytest.ini
plugins: doctestplus-0.11.2, typeguard-2.13.3, xvfb-2.0.0
collecting ... WARNING: Failed to import any usable sherpa.plotting backend. Plotting routines will not be available.
collected 7555 items                                                                                         ```

The report is in ``report/index.html``, which links to individual
files and shows exactly which lines were excuted while running the tests.

Run doctests locally
--------------------
If `doctestplus <https://pypi.org/project/pytest-doctestplus/>` is installed
(and it probably is because it's part of
`sphinx-astropy <https://pypi.org/project/sphinx-astropy/>`,
which is required to build the documentation locally),
examples in the documentation are run automatically.

> This serves two purposes:
> 
>   - It ensure that the examples we give execute and give the expected output.
>     It is confusing to a user to see an example in the documentation that does not
>     work because it contains bugs to calls functions that have been removed from sherpa.
>   - This also acts as an additional tests.
> 
> However, many examples were written because doctestplus was used with Sherpa and do not execute
> as written (e.g. they need implicit set-up that's not spelled out),
> so a number of rst files are explicitly excluded from that run in `pyests.ini`.
> When an rst file is fixed to work top-to-bottom, it should be removed from the
> blacklist in `pytest.ini`.
> 
> doctestplus can also run the examples in the doctrings of individual
> functions, but again many files are still blacklisted for this functionality in
> `pytest.ini`.
> During development, you can run doctestplus on individual files like so::

Reviewer: ```suggestion
This serves two purposes:

  - it ensure that the examples we give are actually correct and match the code,
  - and it acts as additional tests of the Sherpa code base.

The `doctest_norecursedirs` setting in the `pytests.ini` file is used to exclude files which can not be
tested. This is generally because the examples were written before doctestplus support was added, and so
they need to be re-worked, or there is too much extra set-up required that would make the examples
hard-to follow. The file should be removed from this list when it has been updated to allow testing with doctestplus.

During development, you can run doctestplus on individual files like so (the option to use depends on whether it is a Python or reStructuredText file)::
```

Is this better? I feel like we don't need to strongly explain why testing is good.

   pytest --doctest-plus sherpa/astro/data.py
   pytest --doctest-plus sherpa/data.py
   pytest --doctest-rst docs/quick.rst
   pytest --doctest-rst docs/evaluation/combine.rst

> This guarantees that the examples actually work and don't have typos or outdated
> parameters, which might confuse a user.

Reviewer: ```suggestion
```

I think this sentence is superfluous.

If `doctestplus <https://pypi.org/project/pytest-doctestplus/>` is installed
(and it probably is because it's part of
`sphinx-astropy <https://pypi.org/project/sphinx-astropy/>`,
which is required to build the documentation locally),
examples in the documentation are run automatically.
This serves two purposes:

  - It ensure that the examples we give execute and give the expected output.
    It is confusing to a user to see an example in the documentation that does not
    work because it contains bugs to calls functions that have been removed from sherpa.
  - This also acts as an additional tests.

However, many examples were written because doctestplus was used with Sherpa and do not execute
as written (e.g. they need implicit set-up that's not spelled out),
so a number of rst files are explicitly excluded from that run in `pyests.ini`.
When an rst file is fixed to work top-to-bottom, it should be removed from the
blacklist in `pytest.ini`.

doctestplus can also run the examples in the doctrings of individual
functions, but again many files are still blacklisted for this functionality in
`pytest.ini`.

> During development, you can run doctestplus on individual files like so::

Reviewer: This needs some work. Let me suggest updates after the coffee has kicked in and then left my system...
